source/_posts/2016-10-24-gsoc-2016-port-nmatrix-to-jruby.markdown

```diff
@@ -0,0 +1,536 @@
+![Alt Gamma operator](https://github.com/prasunanand/gsoc_blog/blob/master/img/sciruby_blog/plots/gamma.png?raw=true "Fig.6. Gamma Operator")
+![Alt Determinant](https://github.com/prasunanand/gsoc_blog/blob/master/img/sciruby_blog/plots/det.png?raw=true "Fig.7. Determinant")
+![Alt LU Facorization](https://github.com/prasunanand/gsoc_blog/blob/master/img/sciruby_blog/plots/lud.png?raw=true "Fig.8. LU Facorization")
+
+**Result:**
+
+1. For, two dimensional matrices, NMatrix-JRuby is currently slower than NMatrix-MRI for matrix multiplication and matrix decomposition functionalities(calculating determinant and factorizing a matrix). NMatrix-JRuby is faster than NMatrix-MRI for other functionalities of a two dimensional matrix like addition, subtraction, trigonometric operations, etc.
+
+2. NMatrix-JRuby is a clear winner when we are working with matrices of arbitrary dimensions.
+
+
+## **Storing n-dimensional matrices as flat arrays**
+
+The major components of NMatrix are shape, elements, dtype and stype. Any nmatrix when initialized, stores the elements as a flat array. ArrayRealVector class is used to store the elements.
+
+`@s` stores elements, `@shape` stores the shape of the NMatrix, while `@dtype` and `@stype` store the data type and storage type respectively. Currently, we have nmatrix-jruby implemented only for double and Ruby object data type.
+
+
+NMatrix-MRI uses `@s` which is an object containing elements, stride and offset (as in C, we need to deal with the memory allocation for the arrays).
+
+![NMatrix](https://github.com/prasunanand/gsoc_blog/blob/master/img/sciruby_blog/nmatrix.png?raw=true "Fig.1. NMatrix")
+
+## **Slicing and Rank**
+
+Implementing slicing was the toughest part of NMatrix-JRuby implementation.
+`NMatrix@s` stores the elements of a matrix as a flat_array. The elements along any dimension are accessed with the help of the stride. `NMatrix#get_stride` calculates the stride with the help of the dimension and shape and returns an array.
+
+```ruby
+def get_stride(nmatrix)
+  stride = Array.new()
+  (0...nmatrix.dim).each do |i|
+    stride[i] = 1;
+    (i+1...dim).each do |j|
+      stride[i] *= nmatrix.shape[j]
+    end
+  end
+  stride
+end
+```
+
+`NMatrix#[]` and `NMatrix#[]=` are thus able to read and write the elements of a matrix. NMatrix#MRI uses the `@s` object which stores the stride when the nmatrix is initialized.
+
+`NMatrix#[]` calls the `#xslice` operator which calls  `#get_slice` operator that use the stride to determine whether we are accessing a single element or multiple elements. If  there are multiple elements, `#dense_storage_get` returns an NMatrix object with the elements along the dimension.
+
+NMatrix-MRI differs from NMatrix-JRuby implementation as it makes sure that memory is properly utilized as the memory needs to be properly **garbage collected**.
```

Comment:
I am referring to the fact that "JRuby is auto garbage-collected".  What I wrote is misleading.